lua/neogit/lib/git/branch.lua

```diff
@@ -0,0 +1,77 @@
+local a = require('neogit.async')
+local cli = require('neogit.lib.git.cli')
+local util = require('neogit.lib.util')
+local input = require('neogit.lib.input')
+local M = {}
+
+local function contains(table, val)
+   for i=1,#table do
+      if table[i] == val then
+         return true
+      end
+   end
+   return false
+end
+
+local function parse_branches(output)
+  local branches = util.split(output, '\n')
+  local other_branches = {}
+  for _, b in ipairs(branches) do
+    local branch_name = b:match('^  (.+)')
+    if branch_name then table.insert(other_branches, branch_name) end
+  end
+
+  return other_branches
+end
+
+local get_local_branches = a.sync(function ()
+  local branches = a.wait(cli.branch
+    .list
+    .call())
+
+  return parse_branches(branches)
+end)
+
+local get_all_branches = a.sync(function ()
+  local branches = a.wait(cli.branch
+    .list
+    .all
+    .call())
+
+  return parse_branches(branches)
+end)
+
+local function prompt_for_branch(options)
+  local chosen = input.get_user_input_with_completion('branch > ', options)
+  if not contains(options, chosen) then
+    print('ERROR: Creating a new branch from this dialog is not supported (yet)')
+    return
+  end
+  return chosen
+end
+
+M.checkout_local = a.sync(function ()
+  local branches = a.wait(get_local_branches())
+
+  a.wait_for_textlock()
+  local chosen = prompt_for_branch(branches)
+  a.wait(cli.checkout.branch(chosen).call())
+end)
+
+M.checkout = a.sync(function ()
+  local branches = a.wait(get_all_branches())
+
+  a.wait_for_textlock()
+  local chosen = prompt_for_branch(branches)
+  a.wait(cli.checkout.branch(chosen).call())
+end)
+
+M.checkout_new = a.sync(function ()
```

Comment:
Why do you think an extra binding for specifically creating a new branch is good if you could also create a branch by using the other two bindings?

Reply:
Mnemonics for now, mostly. 
I'm also not entirely sure on if we can and want to allow branch creation from the other dialogs as well. For "checkout anything" (`bb`), since it's supposed to checkout anything, we'd need to check against branches, tags and commit hashes if we want to create a branch on other inputs; at that point, it might be easier to just throw it into `git branch ...` and let the error bubble up.
For the distinction between checkout local and create new branch, it's pretty much just the choice of dialog: For checkout local we provide a dialog with completion options, for create new just prompt for any text. Not much of a distinction, but the former could potentially be plugged by telescope, while it doesn't make that much sense for the latter.